IsEditable combox box

almightyju opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
Adds IsEditable property to ComboBox and adds a ItemTextBinding property to get the text value. 

## What is the current behavior?
Currently it's unsupported.

## What is the updated/expected behavior with this PR?
To be able to bind text values in a combo box and it handle the text values and SelectedItem values


## How was the solution implemented (if it's not obvious)?
Added a text box which is bound to TextProperty and only visible when IsEditable is true. Compares the value of the text box with the ItemsSource via an IBinding (similar to the auto complete box)


## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
None


## Fixed issues
#205. 
I am aware that pull request #1070 but it looks like a lot of "ToString" stuff was going on which doesn't really fit with Avalonia, hoping this way is more acceptable

almightyju committed:
IsEditable combox box with text bindings

thevortexcloud commented:
This sounds very similar to the autocomplete box? Is it different somehow?

https://docs.avaloniaui.net/docs/reference/controls/autocompletebox

cla-avalonia commented:
@almightyju,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

almightyju commented:
> This sounds very similar to the autocomplete box? Is it different somehow?
> 
> https://docs.avaloniaui.net/docs/reference/controls/autocompletebox

In 2 ways that I can see:

1. combo box has a button to open the dropdown to show all items
2. when you press backspace in the auto complete box it deletes all text, so it's not a true "editable" experience

It's also a much more familiar experience for people coming from WPF that I think it's worth the small change

almightyju commented:
> @almightyju,
> 
> Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
> 
> ```
> @cla-avalonia agree
> ```
> 
> Contributor License Agreement

@cla-avalonia agree

MrJul commented:
Public API for review:

```diff
namespace Avalonia.Controls
{
    public class ComboBox : SelectingItemsControl
    {
+        public static readonly DirectProperty<ComboBox, bool> IsEditableProperty;
+        public static readonly StyledProperty<string?> TextProperty;
+        public static readonly StyledProperty<IBinding?> ItemTextBindingProperty;
+        public bool IsEditable { get; set; }
+        public string? Text { get; set; }
+        public IBinding? ItemTextBinding { get; set; }
    }
}
```

MrJul commented:
During the API review meeting, we noticed that WPF uses `TextSearch.TextPath` for this purpose instead of `ItemTextBinding`.
In Avalonia, we want a similar API, but compatible with trimming.

Remove:
 - `ComboBox.ItemTextBindingProperty`

Add:
 - `TextSearch.TextBindingProperty` (attached)

Also, `IsEditableProperty` should be a `StyledProperty`.

(Not API related, but `TextSearch.Text` is currently used to search in `ItemsControl`, and should still work with the new implementation.)

almightyju commented:
> During the API review meeting, we noticed that WPF uses `TextSearch.TextPath` for this purpose instead of `ItemTextBinding`. In Avalonia, we want a similar API, but compatible with trimming.
> 
> Remove:
> 
>     * `ComboBox.ItemTextBindingProperty`
> 
> 
> Add:
> 
>     * `TextSearch.TextBindingProperty` (attached)

I wasn't aware of `TextSearch.TextBindingProperty` but I don't see how I can use that in this case, when the text changes it's necessary to loop all of the items and check if the entered text matches an item to set the `SelectedItem` and `SelectedIndex`:

https://github.com/almightyju/Avalonia/blob/2f1a66f467d4aa6c935bf6fd9feb7db70a3201d9/src/Avalonia.Controls/ComboBox.cs#L686-L696

I don't see a way to do that using the attached property since I can't access the `IBinding` to execute against, which is why I made it the same way the auto complete box works.

MrJul commented:
> I wasn't aware of `TextSearch.TextBindingProperty` but I don't see how I can use that in this case, when the text changes it's necessary to loop all of the items and check if the entered text matches an item to set the `SelectedItem` and `SelectedIndex`:

To clarify, `TextSearch.TextBindingProperty` (of type `IBinding`) needs to be added: it doesn't exist right now.
It's really the equivalent of your `ItemTextBinding`, but as an attached property, that could be used for any text search, like in WPF.

I've implemented it in #18405 so this PR can focus on the implementation of `IsEditable` instead.
Once it's merged, you can reuse the `ItemsControl.GetIndexFromTextSearch()` for text searching.

almightyju committed:
Update after #18405 api review

almightyju committed:
ComboBox tests for IsEditable

almightyju commented:
@MrJul thanks for your help with this 😄 I've updated it to work with what you've done and put some tests in (and fixed broken ones).

Also made a docs pull request

thevortexcloud reviewed (commented):
I have left some of my thoughts. But I am not in charge around here, so someone may disagree.

almightyju committed:
remove unneeded error being throw when combox is editable and doesn't have text binding properties

almightyju commented:
Is there anything else I can do to help get this merged?

MrJul committed:
Merge branch 'master' into master

MrJul reviewed (changes requested):
Overall, this seems to work as expected. There are a couple of caveats regarding usability.

Also, when compared to WPF, this PR doesn't provide auto-completion. However, that can be consider another feature and can be implemented later.

MrJul commented on src/Avalonia.Controls/ComboBox.cs:
```diff
@@ -103,6 +121,12 @@ static ComboBox()
+            TextSearch.TextBindingProperty.Changed.AddClassHandler<ComboBox>((x, e) => x.ItemTextBindingChanged(e));
+            //when the items change we need to simulate a text change to validate the text being an item or not and selecting it
+            ItemsSourceProperty.Changed.AddClassHandler<ComboBox>((x, e) => x.TextChanged(
+                new AvaloniaPropertyChangedEventArgs<string?>(e.Sender, TextProperty, x.Text, x.Text, e.Priority)));
```
Indeed, that's not expected, this should be an implementation detail.

almightyju committed:
Changes after code review

almightyju commented:
All done, only 1 thing I can't seem to work out, when focusing backwards (shift + tab) it correctly focuses the "previous" control, but it doesn't have the styling showing like it should, but keypresses do register on the control. 

To demonstrate:
1.  open the control catalog
2. go to combo box page and click in the Ediable combo box (bottom right)
3. press shift tab
4. press down key (notice how it now has Name 1Id 1 in the box)
5. press shift tab again 
6. press tab
7. notice how that combo box is now showing as focused

I have no clue how to fix it...

relevant code: 
https://github.com/almightyju/Avalonia/blob/1abeb9829dc37b71d8277747860f50048599b2aa/src/Avalonia.Controls/ComboBox.cs#L450-L453

almightyju committed:
only do naviagtion check if combo box is editable

MrJul committed:
Fix editable ComboBox tab navigation

MrJul reviewed (changes requested):
Please also handle review comments from @thevortexcloud (which I find pertinent) that are still unresolved.

almightyju committed:
combobox code review improvements

almightyju commented:
> > All done, only 1 thing I can't seem to work out, when focusing backwards (shift + tab) it correctly focuses the "previous" control, but it doesn't have the styling showing like it should, but keypresses do register on the control.
> 
> I've fixed it in commit [3adda9c](https://github.com/AvaloniaUI/Avalonia/commit/3adda9c3d487c12daedbec495443799915a0330c): the trick is set `IsTabStop = false` for the whole `ComboBox` in this case. That's how it's done in `NumericUpDown` in Avalonia, and in `ComboBox` in WPF.
> 
> I've also set `TabNavigation` to `Local` so that tab indexes work as expected in this case.

I knew there would be an easier way, just wasn't sure quite what it was.

I've done the rest of the other requests. Also highlighted an issue I found so added an extra test for when changing the `ItemsSource` with an already set `Text` it wouldn't reselect and would loose the `Text`

MrJul reviewed (approved):
LGTM, thank you!
